Lib/zipfile.py

```diff
@@ -2228,11 +2229,18 @@ def joinpath(self, add):
 
     @staticmethod
     def _add_implied_dirs(names):
-        return names + [
+        subdirs = set([
```

Comment:
This technique is likely to change the order of the names, which seems undesirable to me. Probably better would be to use [unique_everseen](https://docs.python.org/3/library/itertools.html#itertools-recipes) to eliminate duplicates without changing order.

Reply:
Thanks for showing unique_everseen. As I understand, this is a new function I will have to add to zipfile.py, correct? This will also guarantee order.

Reply:
I have updated my PR with a bunch of changes based on your review including using unique_everseen. I am still thinking about the naming issue you brought up for testing the 3 files.